### ISSUE TITLE
README  example two simplifying

### ISSUE THREAD

**Issue by hasansaad1** [foundersandcoders/ws-pure-functions-easy-testing#1]
Simplify example two: remove the closure from updateDom. Simply have a function that does the DOM update based on an input parameter and another function that calculates the value of that parameter based on some other logic.

**Comment by mattlub** [comment]
Hi- I was trying to fix this issue because we delivered the workshop today, and I'm pretty sure the example as it is doesn't work (we skipped the example in the end).

This is what I had after trying to simplify, reading @hasansaad1's PR, and the comments from @eliascodes:

**initial non-pure**
```js
var isVisible = true;

function toggleImageVisibility() {
  var myImage = document.getElementById("myImage");

  if (isVisible === true) {
    myImage.style.opacity = 0;
    isVisible = false;
  }

  else if (isVisible === false) {
    myImage.style.opacity = 1;
    isVisible = true;
  }

}
toggleInvisibility();
```

**refactored purer**
```js
function updateOpacity(element, opacity) {
  element.styles.opacity = opacity;
}

function getNextOpacity(imageIsVisible) {
  if (imageIsVisible === true) {
    return 0;
  } else if (imageIsVisible === false) {
    return 1;
  }
}

var isVisible = true

function toggleInvisibility() {
  var myImage = document.getElementById('myImage')
  updateOpacity(myImage, getNextOpacity(isVisible))
  isVisible = !isVisible
}

// When we're ready...
toggleInvisibility();
```

However this doesn't seem much better, as `updateOpacity` is the only real improvement to the  `toggleInvisibility` function, and it is not even pure.

-----------

Any thoughts? If someone has a nicer example I'd be happy to try to incorporate it into the README instead of the current one.

**Comment by eliasmalik** [comment]
This example possibly suffers from the fact that the example of a toggle is _too_ simple. The logic is really just `s => !s`. Everything else needs to interact with the DOM API to some degree.

It may be clearer to learners if the functionality being demonstrated had a more substantial logic component. At the same time, it's easy to go too far here, and make the example incomprehensible.

**Comment by mattlub** [comment]
@eliascodes I agree, but currently I think the example _is_ incomprehensible. Do you know what has happened when this has been delivered in London?

**Comment by eliasmalik** [comment]
> Do you know what has happened when this has been delivered in London?

Not really, better off asking the CFs.

What about cycling instead of toggling? `A -> B -> C -> A ...`

Before:
```js
(() => {
  const node = document.querySelector('...');

  if (node.classList.contains('A')) {
    node.classList.replace('A', 'B');
  } else if (node.classList.contains('B')) {
    node.classList.replace('B', 'C');
  } else if (node.classList.contains('C')) {
    node.classList.replace('C', 'A');
  } else {
    node.classList.add('A');
  }
})()
```

After:
```js
const getNextClass = (class, list) => 
  list[(list.indexOf(class) + 1) % list.length];

const replaceClass = (old, new, node) =>
  node.classList.replace(old, new);

// Usage:
const classes = ['A', 'B', 'C'];
let current = classes[0];
(() => {
  const node = document.querySelector('...');
  
  // Calculate state (pure)
  const next = getNextClass(current, classes);

  // Effect
  replaceClass(current, next, node);
})()
```

No need to do `getNextClass` in a one liner.

**Comment by mattlub** [comment]
re my comment above: @matthewdking @shiryz @arrested-developer @yasminhillis, do you remember anything about when this was last delivered in other campuses?

Anyway, I'm happy to make a PR to update the example to what Elias suggested above, which I think is better.

**Comment by shiryz** [comment]
@mattlub I don't recall any problems from the last time it was delivered, checked the [sgc minutes](https://github.com/foundersandcoders/nazareth-programme/blob/master/stop-go-continue/facn4/week2.md#schedule) and it seems to be fine as well...
I would say though that Elias's example is really abstract and something with actual classes/names would make more sense :upside_down_face: (which I'm pretty sure is what you were going for :smiley:)

**Comment by arrested-developer** [comment]
Feedback from FAC15 a couple of weeks ago:

### Pure functions workshop: 
- mistakes in readme, its talking about stuff that doesnt exist in the code
- people not sure whats the point of the exercises, the descriptions should be clearer, so they know what they're expected to do
- most did the first 3 functions quite quickly but were rather stuck on the increment array and the following ones
- the tests when failing say "the original array had not been altered" which people understood meant they should be mutating the original array in order to pass test.

**Comment by zurda** [comment]
We delivered this last week with FAC16 and we didn't notice any unusual confusion.
